torch/cuda/memory.py

```diff
@@ -590,3 +606,110 @@ def mem_get_info(device: Union[Device, int] = None) -> int:
         device = torch.cuda.current_device()
     device = _get_device_index(device)
     return torch.cuda.cudart().cudaMemGetInfo(device)
+
+
+def set_enabled_uvm(enable):
+    r"""Enable/disable Unified Virtual Memory.
+    Arguments:
+        enable (bool): desired UVM setting.
+    """
+    torch._C._cuda_setEnabledUVM(enable)
+
+
+def get_enabled_uvm():
+    r"""Returns a bool indicating if Unified Virtual Memory is currently enabled."""
+    return torch._C._cuda_getEnabledUVM()
+
+
+def set_enabled_move(enable):
+    r"""Enable/disable Unified Virtual Memory.
+    Arguments:
+        enable (bool): desired UVM setting.
+    """
+    torch._C._cuda_setEnabledMove(enable)
+
+
+def get_enabled_move():
+    r"""Returns a bool indicating if Unified Virtual Memory is currently enabled."""
```

Comment:
Thanks. Please let me add comments here.

What's the difference between "UVM" and "Move"? The description looks the same.

Reply:
Hi @shintaro-iwasaki !

Enabling "UVM" will cause all getAllocator calls to return the new Caching Managed Allocator.
Enabling "Move" will attempt to short-circuit copy calls that are determined to only be changing the device (as opposed to other tensor transformations, which would require a copy)

Reply:
The idea with enabling the move is deferring to the UVM driver to handle data placement. With Moves off, it will always copy, even if UVM is on - from managed alloc to managed alloc.

Reply:
Thanks for your reply. So do you mean the following?
- When "UVM" is on, the new UVM allocator will be used.
- When "UVM" is off, the existing memory allocator will be used.
- When "Move" is on, PyTorch checks if data is on UVM. If it's on UVM, the data is not explicitly copied. If it's not on UVM, the data is explicitly copied.
- When "Move" is off, PyTorch always explicitly copies data without check if data is on UVM or not.

Then when do you want to disable "Move"? Is this check so heavy? Or is there any performance implication?

Reply:
Almost.

>>When "Move" is on, PyTorch checks if data is on UVM. If it's on UVM, the data is not explicitly copied. If it's not on UVM, the data is explicitly copied.

When Move is on, Pytorch does a few different checks before allowing the UVM driver control:
 - Is UVM enabled?
 - Is the storage ptr for the Tensor on a managed allocation?
 - Has the device changed in this to() command?
 - Are the other parameters unchanged? (value, layout, mem format)

These checks are in the `aten/src/ATen/native/TensorConversions.cpp` file (in `to_will_move()` and `_to_move`

Reply:
We made move a separate enable/disable so that we could test each part separately, mostly.
There could be a scenario, however, where you know allowing the UVM driver to do copies will be slower but you still care about having tensors on managed memory to avoid running out of GPU memory.

Reply:
> allowing the UVM driver to do copies will be slower but you still care about having tensors on managed memory to avoid running out of GPU memory.

i'm confused about how you could take advantage of managed memory to avoid running out of GPU memory but also avoid slow copies.  Could you elaborate on which type of system you're referring to and whether the cure to running out of memory relates to 'move' functionality?  Is 'move' enablement orthogonal to the functionality in the driver that pages data between gpu/cpu seamlessly as needed by gpu?

Reply:
>> the cure to running out of memory relates to 'move' functionality?
No, that's not what I meant. Using managed allocations will help with running out of memory. 

>> Is 'move' enablement orthogonal to the functionality in the driver that pages data between gpu/cpu seamlessly as needed by gpu?

yes. The driver can page in/out based on need. (more memory is needed, or a chunk is being accessed)
Move is a function to defer Pytorch's explict copying to the UVM driver when enabled. When off Pytorch will copy from one managed chunk to another (like normal) where the target managed chunk was hinted to be on GPU. 

On dGPUs using Move may be a bit slower. In this case the data does need to get to GPU memory and generally asap. In the move kernel we do a prefetch using that target device instead of an explicit copy and it's up to the driver to copy. The UVM driver also copies on page boundaries which could be more than needed. That being said prefetching managed chunks can be tuned, using events and hints, to become close performance-wise to explicit copies. Nvidia has a blog showing some of this: https://developer.nvidia.com/blog/maximizing-unified-memory-performance-cuda/

There is room for more tuning in the Move kernel that's here.